extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,375 @@
+ *    element, at which point it will return to the INITIAL state.
+ *
+ * @enum {string}
+ */
+const RefreshLifecycleState = {
+  /**
+   * Element has been registered, but not yet seen on screen.
+   */
+  INITIAL: 'initial',
+
+  /**
+   * The element has appeared in the viewport, but not yet for the required
+   * duration.
+   */
+  VIEW_PENDING: 'view_pending',
+
+  /**
+   * The element has been in the viewport for the required duration; the
+   * refresh interval for the element has begun.
+   */
+  REFRESH_PENDING: 'refresh_pending',
+
+  /**
+   * The refresh interval has elapsed; the element is in the process of being
+   * refreshed.
+   */
+  REFRESHED: 'refreshed',
+};
+
+/**
+ * An object containing the IntersectionObservers used to monitor elements.
+ * Each IO is configured to a different threshold, and all elements that
+ * share the same visiblePercentageMin will be monitored by the same IO.
+ *
+ * @const {!Object<string, (!IntersectionObserver|!IntersectionObserverPolyfill)>}
+ */
+const observers = {};
+
+/**
+ * An object containing all currently active RefreshManagers. This is used in
+ * the IntersectionOberserver callback function to find the appropriate element
+ * target.
+ *
+ * @const {!Object<string, !RefreshManager>}
+ */
+const managers = {};
+
+/**
+ * Used to generate unique IDs for each RefreshManager.
+ * @type {number}
+ */
+let refreshManagerIdCounter = 0;
+
+export class RefreshManager {
+
+  /**
+   * @param {!./amp-a4a.AmpA4A} a4a The AmpA4A instance to be refreshed.
+   * @param {!RefreshConfig} config
+   */
+  constructor(a4a, config) {
+
+    /** @private {string} */
+    this.state_ = RefreshLifecycleState.INITIAL;
+
+    /** @const @private {!./amp-a4a.AmpA4A} */
+    this.a4a_ = a4a;
+
+    /** @const @private {!Window} */
+    this.win_ = a4a.win;
+
+    /** @const @private {!Element} */
+    this.element_ = a4a.element;
+
+    /** @const @private {string} */
+    this.adType_ = this.element_.getAttribute('type').toLowerCase();
+
+    /** @const @private {?number} */
+    this.refreshInterval_ = this.getPublisherSpecifiedRefreshInterval_();
+
+    /** @const @private {!RefreshConfig} */
+    this.config_ = this.convertConfiguration_(config);
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = timerFor(this.win_);
+
+    /** @private {?(number|string)} */
+    this.refreshTimeoutId_ = null;
+
+    /** @private {?(number|string)} */
+    this.visibilityTimeoutId_ = null;
+
+    /** @private {boolean} */
+    this.isRefreshable_ = isExperimentOn(this.win_, 'amp-ad-refresh') &&
+        // The network has opted-in.
+        !!(this.config_ &&
+        // The publisher has enabled refresh on this slot.
+        (this.refreshInterval_ || this.refreshInterval_ == '') &&
+        // The slot is contained only within container types eligible for
+        // refresh.
+        !getEnclosingContainerTypes(this.element_).filter(container =>
```

Comment:
arguably this constraint should be within doubleclick impl and not refresh manager as other networks will likely not have the same constraint.  Mind moving it?